### PR TITLE
New MSMonitoring module

### LINKED ISSUE #11429: [MS] Create monitoring dashboard for MSPileup microservice
**Impact of the new feature**
New WMCore microservice

**Is your feature request related to a problem? Please describe.**
Creation of a new microservice to make pileup management more efficient and easier.
Subtask of https://github.com/dmwm/WMCore/issues/9779

**Describe the solution you'd like**
This is a joint ticket between the WMCore and the CMS Monitoring team, such that we can define and provide a monitoring dashboard for the MSPileup. So, relevant JIRA ticket might have to be created to get the usual process monitoring, as we have for the other WMCore microservices.

**Describe alternatives you've considered**
We can eventually consider service specific metrics as well, but I'd call that to be an future enhancement and we can properly define that in the future.

**Additional context**
None

### PR DESCRIPTION
**Pull request by vkuznet**
Fixes #11429 

#### Status
In development

#### Description
New MSMonitoring module provides everything we need to perform monitoring task for MSPileup. The new class can be initialized with AMQ credentials, and provide functions to flatten MSPileup docs and send them over to CMS Monitoring AMQ end-point.

The wiki documentation is update over [here](https://github.com/dmwm/WMCore/wiki/ReqMgr2-MicroService-Pileup#cms-monit-task)

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
The corresponding changes to config can be found in this PR: https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/196

### CODE REVIEW

**Comment by amaltaro** [comment]
@vkuznet Valentin, before you continue working on this, note that all microservices define a cherrypy thread performing the service monitoring. It involves
a) fetch and upload data to wmstats
b) can potentially fetch and upload data to AMQ as well

Relevant service section can be found here:
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/blob/preprod/reqmgr2ms-pileup/config.py#L93

**Review by amaltaro — COMMENTED** [review]
Please see comment above.

**Comment by vkuznet** [comment]
Alan, this is a first draft of MSPileupMonitoring module which contains initial implementation of the logic. I need an initial review of the functionality in order to polish it further. Few things to note:
- I decided to not use `CherryPy` thread as this module can run without it
- I re-used `daemon` from `MSManager` where I added `THR_FLAG` to allow restart of the thread and ability to gracefully exit it. I suggest to put this functionality back to `MSManager` though but I decided to give you heads-up to review the implementation
- I re-used `uploadToAMQ` method from `src/python/WMCore/REST/HeartbeatMonitorBase.py` module and I think if we'll have an intermediate class for AMQ, e.g. CMSMonitoring, then we can put it over there to share across different classes, e.g. make `CMSMonitoring` as a based class for them which will have this method. But so far I put it to `MSPileupMonitoring`.

Please make an initial review and provide your feedback, then I can proceed with this PR.

**Comment by vkuznet** [comment]
I strongly believe that monitoring task should not be tighten to CherryPy thread. Please review and decide, if you can convince me with real arguments why thread should be in CherryPy and what are the benefits I can put it over there. So far it only make WMCore depends deeper and deeper on Cherrypy for tasks which are not even remotely related to HTTP activity, e.g. Monitoring has nothing to do with HTTP (which CherryPy provides).

**Comment by amaltaro** [comment]
You are confusing things here.

A cherrypy thread is simply a mechanism to create another task (thread) that runs along with your service, where we define a specific interval to get that task executed. That is exactly what we need to have for this monitoring purpose, something that is probing the service and collecting monitoring information every amount of time.

I do acknowledge one pitfall on this implementation though, which is, if the service is down so is the monitoring thread. However, for that we have already a few other mechanism to either spot it or to restart the service (liveness probe, wmstats grace period).

If we move away of CherryPy and the future web framework does not provide a similar functionality, then yes, we need to have one specific service/thread for the monitoring purpose. For now, I see no strong reason why we should not use what is already used by all of the other services and that serves all our purposes.

**Comment by vkuznet** [comment]
Alan, I provided my feedback on implementation but I'll not insist on using standard library approach. If you feel comfortable with using cherrypy thread for that I can easily convert the code to use it. For that I need the following pieces:
- currently `WMCore/MicroService/CherryPyThreads/HeartbeatMonitor.py` is delegating its job to `HeartbeatMonitorBase` from `WMCore.REST.HeartbeatMonitorBase` module, but
- it is unclear who collect the data. I need to understand this since for MSPileup we need to fetch data from its backend. I implemented the appropriate method in this PR. So, you need to provide me concrete example how other MS service collect their data (if any) and send it over to specific destination (wmstats or CMS Monitoring). From what I inspect in a code I do not see that this is done anyhow.
- I also do not have enough details to understand why do we need to send data to wmstats, please elaborate more on this, e.g. which data and in which data-format we should collect and send to  wmstats. In my view, we have robust CMS Monitoring infrastructure which can handle any data suited for monitoring. Why we should involve wmstats in monitoring tasks it is unclear to me.

Once I'll know all aforementioned details I can move forward with this PR.

**Comment by amaltaro** [comment]
Valentin, given that we already have the basic monitoring performed by the CherryPy thread and uploaded to WMStats, with this module:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/REST/HeartbeatMonitorBase.py

I would suggest to put this development on hold and tackle other tickets first. We will likely need the development already done here though, because I fail to see how to add extra monitoring and push data to AMQ without further development in the microservices.

I think we now have all the pieces in place and we can start focusing on the actual algorithm for this microservice, to be addressed with the following 2 tickets:
https://github.com/dmwm/WMCore/issues/11426
https://github.com/dmwm/WMCore/issues/11428

Once those are covered, we can then get back to the auth/authz issue, the dashboard monitoring, and this extra/custom monitoring ticket.

**Review by amaltaro — CHANGES_REQUESTED** [review]
Valentin, in addition to the comments made along the code, could you please also provide an extra commit enabling this monitoring thread in MSPileup? I assume you are going to use either MSPileup or MSPileupTasks for that?

Once that is made, we also need to check whether there won't be any conflicts with the cherrypy thread that can be potentially updating data to AMQ. I think we will be fine, given that the cherrypy thread shouldn't have any documents to be uploaded. But it's worth it a double check.

File: src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py
Comment (vkuznet) [reply src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py:49]: Alan, apart from doc string typo, I do not understand what you mean "wrong". I decided to flatten the document from and replace keys with plural name to single, e.g. "ruleIds" which is a list will become "ruleId" which will have single value. I tested this functionality in provided unit test. This will correctly reflect the data-type, since "rulesIds" is a list data type, while when we will flatten the dict we will send only one element of such list, i.e. in this case a string. I just need to understand what you meant by "wrong" and I can adjust accordingly.

Please note, to simplify queries in ES and grafana dashboards, it is desired to flatten the document. Since we have few lists, I thought it will be natural to use plural to single conversion, e.g. "ruleIds" will become "ruleId, "currentRSEs" will become "currentRSE", etc.

File: src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py
Comment (vkuznet) [reply src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py:36]: done

File: src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py
Comment (vkuznet) [reply src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py]: done

File: src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py
Comment (amaltaro) [reply src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py:50]: Ahh, now I understand what you have done in the code. Yes, you are right.
It might be better to add a comment in the line above though saying something like:
```
# convert plural to singular key name
```

About the docstring, please read: https://github.com/dmwm/WMCore/blob/master/CONTRIBUTING.rst#project-docstrings-best-practices and adjust the code accordingly.

File: src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py
Comment (amaltaro) [reply src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py:56]: I guess I need to have another look at this logic then. But in general, relying on a global variable isn't a good start. Sorry, I cannot suggest any other alternative at the moment.

Comment (vkuznet) [reply src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py:56]: I agree about globals and I also don't like them. The only alternative I see is to use signal, something like this:
```
def signal_handler(sig, frame):
    print('You pressed Ctrl+C!')

signal.signal(signal.SIGINT, signal_handler)
```
I need to experiment with this but it may allow to replace global with signal handler. Do you want me to try it out?

File: src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py
Comment (vkuznet) [reply src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py:50]: ok

**Review by amaltaro — COMMENTED** [review]
Valentin, you might have missed my previous comment, so I paste it here again:
"""
Valentin, in addition to the comments made along the code, could you please also provide an extra commit enabling this monitoring thread in MSPileup? I assume you are going to use either MSPileup or MSPileupTasks for that?

Once that is made, we also need to check whether there won't be any conflicts with the cherrypy thread that can be potentially updating data to AMQ. I think we will be fine, given that the cherrypy thread shouldn't have any documents to be uploaded. But it's worth it a double check.
"""

On top of that, can you please update the MSPileup wiki with a brief explanation of the data structure that we have decided to post to AMQ? I understand it's exactly the same as MSPileupObj, but with list values distributed in separate docs. Still worth it to have it there. Thanks

**Comment by vkuznet** [comment]
Alan, I did not forgot about your original request, but it requires merge of this PR https://github.com/dmwm/WMCore/pull/11517 otherwise we'll get conflict in MSPileupTaskManager. Once you'll merge clean-up PR, then I can add requested features over here once I'll rebase the branch.

**Comment by vkuznet** [comment]
@amaltaro , please review this (only this) commit https://github.com/dmwm/WMCore/pull/11478/commits/93bec1154379a4d9d828c1e3384e81eb2e0841f6 It includes a first way how we may integrate MSpileupMonitoring into MSPileupTasks. It will be executed within each cycle and it will simplify MSPileupMonitoring code a lot (I can easily remove all daemon stuff in it). Another integration option will be to start MSPileupMonitoring as separate thread and run it along with MSPileup service. Please evaluate both options and let me know how to proceed. So far I'm in favor the first one, i.e. the one I just committed, since it will be yet another task and does not require additional thread management (and resources).

**Comment by amaltaro** [comment]
> It will be executed within each cycle and it will simplify MSPileupMonitoring code a lot (I can easily remove all daemon stuff in it).

@vkuznet Valentin, I like this idea. This monitoring will be fairly static and there is no need to be monitoring pileups in any aggressive way. I made a couple of comments directly in that commit that you shared, hopefully you can see it.

But yes, just execute this AMQ monitoring task together with the other MSPileupTasks. Thanks

**Comment by vkuznet** [comment]
ok, then, I'll start cleaning up the code and we'll use it as yet another task (of course wiki will require the update too :)

**Comment by vkuznet** [comment]
Alan, please proceed with review. I adjusted code according to our recent zoom discussion. The corresponding PR for configuration (preprod branch) can be found over here: https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/196

File: test/python/WMCore_t/MicroService_t/MSPileup_t/MSPileupMonitoring_t.py
Comment (amaltaro) [inline test/python/WMCore_t/MicroService_t/MSPileup_t/MSPileupMonitoring_t.py:34]: This is not really testing MSPileupMonitoring class, but the utilitarian functions.
As such, I would suggest two minor changes here:
1) give a more meaningful test name, e.g.: `testFlatKeys`
2) whenever possible, create unit tests that are as small as possible. In this case, I would suggest to actually have 2 unit tests, one for each function.

File: src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py:98]: this line should probably be updated.

File: src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py:10]: Given that datetime usage is very basic in this module, why not to simply stick to the `time` module that is already imported here?

File: src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py:64]: It looks like `monitoring_interval` is actually not used. Should we remove it?

File: src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py:75]: how about returning an empty dict as default?

File: src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py
Comment (amaltaro) [inline src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py:107]: Isn't success actually: notifications - failures?

**Review by amaltaro — CHANGES_REQUESTED** [review]
@vkuznet Valentin, please some extra comments along the code.
Once you are done with changes, please ask for another review. If things are ready from your side, please also remove the GH labels (not merge, work in progress).

File: src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py
Comment (vkuznet) [reply src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py:64]: done

File: src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py
Comment (vkuznet) [reply src/python/WMCore/MicroService/MSPileup/MSPileupMonitoring.py:75]: done

File: test/python/WMCore_t/MicroService_t/MSPileup_t/MSPileupMonitoring_t.py
Comment (vkuznet) [reply test/python/WMCore_t/MicroService_t/MSPileup_t/MSPileupMonitoring_t.py:34]: done

**Comment by vkuznet** [comment]
Alan, during squash I made a mistake which put commits of code into commits of unit tests and unfortunately I do not know how to revert such thing (after the squash is done). To avoid further confusion, I squashed everything into single commit and let you know explicitly about it.

Said that, the code is ready for another review.

**Review by amaltaro — APPROVED** [review]
Thanks Valentin, it looks all good to me. I am going to merge this and the services_config PR and cut a new release.

**Pull request merged**
